File: 第三章_字符串/拓扑结构相同子树练习题/IdenticalTree.java
import java.util.*;
 
/*
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;
    public TreeNode(int val) {
        this.val = val;
    }
}*/
public class IdenticalTree {
    public boolean chkIdentical(TreeNode A, TreeNode B) {
        // write code here
        StringBuilder builder=new StringBuilder();
        String s = convertTreeToStr(A, builder);
        builder.delete(0,builder.length());
        String s1 = convertTreeToStr(B, builder);
        return kmp(s, s1) != -1;//判断在A串中是否有B串存在
    }
    //kmp算法 具体详解请看http://www.jianshu.com/p/826480d851ee
    private int kmp(String a,String b){
        int a_len=a.length();
        int b_len=b.length();
        int i=0,j=0;
        int [] next=next(b);
        while (i<a_len&&j<b_len){
            if (j==-1||a.charAt(i)==b.charAt(j)){
               i++;
               j++;
            }else {
                j=next[j];
            }
        }
        if (j==b_len){
            return i-j;
        }else {
            return -1;
        }
    }
    //生成next数组
    private int [] next(String b){
        int b_len=b.length();
        int j=0,k=-1;
        int [] next=new int[b_len];
        next[0]=-1;
        while (j<b_len-1){
            if (k==-1||b.charAt(j)==b.charAt(k)){
                k++;
                j++;
                if (b.charAt(j)!=b.charAt(k)){
                    next[j]=k;
                }else {
                    next[j]=next[k];
                }
            }else {
                k=next[k];
            }
        }
        return next;
    }
    //序列化字符串
    public String convertTreeToStr(TreeNode head,StringBuilder builder){
        Stack<TreeNode> stack=new Stack<>();
        stack.push(head);
        while (!stack.isEmpty()){
            TreeNode poll = stack.pop();
            builder.append(poll.val+"!");
 
            if (poll.right!=null){
                stack.push(poll.right);
            }else {
                builder.append("#!");
            }
 
            if (poll.left!=null){
                stack.push(poll.left);
            }else {
                builder.append("#!");
            }
 
        }
        return builder.toString();
    }
}

File: 第三章_字符串/字符串(1)/拓扑结构相同子树练习题/IdenticalTree.cpp
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};*/

class IdenticalTree {
public:
    bool chkIdentical(TreeNode* A, TreeNode* B) {
        string strA;
        Serialization(A,strA);
        string strB;
        Serialization(B,strB);
 		//判断在A串中是否有B串存在
        int res=kmpSearch(strA,strA.size(),strB,strB.size());
 
        return res!=-1;
    }
	//序列化字符串
	string Serialization(TreeNode *A,string &s){
		if(A==NULL){
			s+="#!";
			return s;
		}

		char ch=A->val+'0';
		s+=ch;
		s+='!';
		
		Serialization(A->left,s);
		Serialization(A->right,s);

		return s;
	}
	
	//kmp算法 具体详解请看http://www.jianshu.com/p/826480d851ee
	int kmpSearch(string A,int lena,string B,int lenb){
        int i=0,j=0;
        int *next=generateNext(B,lenb);
        while(i<lena&&j<lenb)
        {
            if(j==-1||A.at(i)==B.at(j))
            {
                i++;
                j++;
            }
            else
            {
                j=next[j];
            }
        }
        if(j==lenb)
        {
            return i-j;
        }
        else
        {
            return -1;
        }
    }
	
	//生成next数组
	int * generateNext(string B,int lenb)
    {
        int j,k;
        j=0;
        k=-1;
        int * next=new int [lenb];
        next[0]=-1;
        while(j<lenb-1)
        {
            if(k==-1||B.at(k)==B.at(j))
            {
                j++;
                k++;
                if(B.at(k)!=B.at(j))
                {
                    next[j]=k;
                }
                else
                {
                    next[j]=next[k];
                }
            }
            else
            {
                k=next[k];
            }
        }
 
        return next;
    }
};